url5583f3dbb2_XOR19_XScript_tgz-pJ8-xscript_interactive_MainJ8/src/xscript/compiler/scopes/XScope.java

```diff
@@ -0,0 +1,339 @@
+package xscript.compiler.scopes;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import xscript.compiler.XClosureVar;
+import xscript.compiler.XGlobal;
+import xscript.compiler.XJump;
+import xscript.compiler.XJumpTarget;
+import xscript.compiler.XLabel;
+import xscript.compiler.XVar;
+import xscript.compiler.XVarAccess;
+import xscript.compiler.scopes.XScope.Result.R;
+import xscript.compiler.tree.XTree;
+
+public class XScope {
+
+	protected final XModuleScope module;
+
+	protected final XBaseScope base;
+
+	protected final XScope parent;
+
+	protected final XClassScope lastClassScope;
+
+	protected final Map<String, XVar> locals;
+
+	protected final Map<String, XVar> accessed;
+
+	protected final boolean localsAllowed;
+
+	protected final XJumpTarget _break;
+
+	protected final XJumpTarget _continue;
+
+	protected final List<XLabel> labels;
+
+	protected final int pops;
+
+	protected final boolean breakOnlyByName;
+
+	XScope(XScope thiz) {
+		if (thiz.parent == null) {
+			parent = null;
+			this.module = (XModuleScope) this;
+		} else {
+			parent = thiz.parent.lock();
+			this.module = parent.module;
+		}
+		if (this instanceof XBaseScope) {
+			base = (XBaseScope) this;
+			lastClassScope = null;
+		} else {
+			base = parent.base;
+			if (this instanceof XClassScope) {
+				lastClassScope = (XClassScope) this;
+			} else {
+				lastClassScope = parent.lastClassScope;
+			}
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 52:62, 77:87, 101:111, 125:135.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.